File: LeetCode/DPMultiDimension/python/maxProfitIII.py
from typing import List


class Solution:
    def maxProfitMem(self, prices: List[int]) -> int:
        # This is a 3d array mem[2][3][prices.length]
        # mem[true/false][bought/sold/skip][stock positions]
        mem = [[[-1 for _ in range(len(prices))] for _ in range(3)] for _ in range(2)]

        def recursion(prices, pos, t, bought):
            if pos >= len(prices) or t == 0:
                return 0
            if mem[bought][t][pos] != -1:
                return mem[bought][t][pos]

            # 3 choices for a position-> buy/sell/skip
            result = recursion(prices, pos+1, t, bought) # skip
            if bought:
                result = max(result, recursion(prices, pos+1, t-1, False) + prices[pos]) # Sell
            else:
                result = max(result, recursion(prices, pos+1, t, True) - prices[pos]) # Buy

            mem[bought][t][pos] = result

            return result

        # print(mem)
        res = recursion(prices, 0, 2, False)

        return res    

if __name__ == "__main__":
    obj = Solution()

    input1 = [3,3,5,0,0,3,1,4]
    expected1 = 6
    result1 = obj.maxProfitMem(input1)
    print("Input: {}".format(input1))
    print("Expected: {}".format(expected1))
    print("Result: {}".format(result1))
    print("-" * (50))

    input2 = [1,2,3,4,5]
    expected2 = 4
    result2 = obj.maxProfitMem(input2)
    print("Input: {}".format(input2))
    print("Expected: {}".format(expected2))
    print("Result: {}".format(result2))
    print("-" * (50))

    input3 = [7,6,4,3,1]
    expected3 = 0
    result3 = obj.maxProfitMem(input3)
    print("Input: {}".format(input3))
    print("Expected: {}".format(expected3))
    print("Result: {}".format(result3))
    print("-" * (50))

File: LeetCode/DPMultiDimension/java/maxProfitIII.java
package LeetCode.DPMultiDimension.java;

import java.util.Arrays;
import java.util.List;
import java.util.stream.Collectors;
import java.util.stream.IntStream;

class maxProfitIII {
    private List<List<List<Integer>>> mem;

    private int recursion(int[] prices, int pos, int t, int bought) {
        if (pos >= prices.length || t == 0) {
            return 0;
        }
        if (mem.get(bought).get(t).get(pos) != -1) {
            return mem.get(bought).get(t).get(pos);
        }

        // 3 choices for a position-->buy/sell/skip
        int result = recursion(prices, pos + 1, t, bought); // skip
        if (bought == 1) {
            result = Math.max(result, recursion(prices, pos + 1, t - 1, 0) + prices[pos]); // sell
        } else {
            result = Math.max(result, recursion(prices, pos + 1, t, 1) - prices[pos]); // buy
        }

        mem.get(bought).get(t).set(pos, result);

        return result;
    }

    public int maxProfit(int[] prices) {
        mem = IntStream.range(0, 2)
                .mapToObj(x -> IntStream.range(0, 3)
                        .mapToObj(y -> IntStream.range(0, prices.length)
                                .mapToObj(z -> -1)
                                .collect(Collectors.toList()))
                        .collect(Collectors.toList()))
                .collect(Collectors.toList());

        // mem.forEach(x -> {
        // x.forEach(y -> {
        // System.out.println(y);
        // });
        // });

        int res = recursion(prices, 0, 2, 0);
        return res;
    }

    public static void main(String[] args) {
        maxProfitIII soln = new maxProfitIII();

        int[] input1 = new int[] { 3, 3, 5, 0, 0, 3, 1, 4 };
        int expected1 = 6;
        int result1 = soln.maxProfit(input1);
        System.out.println("Input: " + Arrays.toString(input1));
        System.out.println("Expected: " + expected1);
        System.out.println("Result: " + result1);
        System.out.println("-".repeat(50));

        int[] input2 = new int[] { 1, 2, 3, 4, 5 };
        int expected2 = 4;
        int result2 = soln.maxProfit(input2);
        System.out.println("Input: " + Arrays.toString(input2));
        System.out.println("Expected: " + expected2);
        System.out.println("Result: " + result2);
        System.out.println("-".repeat(50));

        int[] input3 = new int[] { 7, 6, 4, 3, 1 };
        int expected3 = 0;
        int result3 = soln.maxProfit(input3);
        System.out.println("Input: " + Arrays.toString(input3));
        System.out.println("Expected: " + expected3);
        System.out.println("Result: " + result3);
        System.out.println("-".repeat(50));
    }
}